### ISSUE TITLE
How to set coef for mmDS

### ISSUE THREAD

**Issue by yuyingxie** [HelenaLC/muscat#55]
I have the following warning message

'Argument 'coef' not specified; testing for “group_idSevere”.'

For this data, group_id inludes 'Healthy', 'Moderate', and 'Severe'. How can I set coef to 'Healthy'?

I tried ' mm <- mmDS(B1, method = "nbinom", n_cells = 10, n_samples = 2,  min_count = 1, min_cells = 20, coef = 'Healthy')'
but had the following error
'Error in .check_args_mmDS(as.list(environment())) :
  is.numeric(u$coef) & u$coef %in% seq_len(nlevels(u$x$group_id)) |  .... is not TRUE'

Why do we need coef to be numeric? It should be categorical, right?

**Comment by plger** [comment]
The correct way to refer to the coefficient would be "group_idHealthy", assuming that it's not the base (i.e. first) level of the factor. You can see the coefficients with `colnames(model.matrix(~group_id, data=as.data.frame(colData(B1))))`, and they can be referred to with numbers in this order.
This being said, I'm not sure you really want to see the effect of "Healthy". If you want to test for genes with differences in any group, the simplest way is probably to set Healthy as the first factor (using `relevel`) and then set the other two coefficients in the `coef` argument of mmDS/pbDS.

**Comment by lindsaynhayes** [comment]
`mmDS` is using a `~group_id + (1|sample_id)` design matrix because it is a mixed effect model right? So it should be 
`colnames(model.matrix(~group_id+(1|sample_id), data=as.data.frame(colData(sce))))` to get the design factors to specify the coef. 

When I run that 

```
colnames(model.matrix(~group_id+(1|sample_id), data=as.data.frame(colData(sce))))

[1] "(Intercept)"       "group_idCoefB"   "group_idCoefC  
[4] "group_idCoefD"   "1 | sample_idTRUE"

Warning message:
In Ops.factor(1, sample_id) : ‘|’ not meaningful for factors
```
Is that a warning message a problem that sample_id is a factor? I tried converting it to character and it really didn't like that either.

**Comment by plger** [comment]
Hi,
under the hood it is, but this is essentially added by `mmDD` to the model you specify. I'm not sure but I don't think model.matrix can handle random effect terms, I had just written it here for the purpose of just figuring out how the other coefficients get named. Since you can't anyway test for the random effect, you might as well just run
```
colnames(model.matrix(~group_id, data=as.data.frame(colData(sce))))
```
to know that.

**Comment by kanefos** [comment]
Hi Pierre-Luc,
Sorry I'm just reading the other documentation and your reply here. Does
> Since you can't anyway test for the random effect

Then mean `mmDS()` doesn't include `sample_id` as a random effect? Or that this function does internally when I call `coef` as the last level of `group_id` ? Is the inability of `mode.matrix()` to handle random effects why `mmDS()` doesn't take a model/contrast matrix as input?

**Comment by plger** [comment]
No, my point was that standard significance testing for mixed models does not test the significance of random effects.
In other words:
```{r}
coef(summary(lmer(y~(1|sample)+x)))
```
doesn't perform any testing on the sample coefficients: 
```
               Estimate Std. Error       df    t value  Pr(>|t|)
(Intercept) -0.17982987 0.21614089 34.74277 -0.8320030 0.4110897
x            0.03302639 0.03371667 96.13796  0.9795271 0.3297781
```
This doesn't mean however that the term is not used. Likewise, `mmDS` internally appends the random term to the formula and uses it in the fit, although it cannot return significance values for it.

It's definitely possible to use contrasts with LMMs, but not, right now, in `mmDS`. (But most questions/tests can anyway be framed either way)
